src/or/connection.c

```diff
@@ -2659,16 +2688,31 @@ retry_listener_ports(smartlist_t *old_conns,
           break;
         }
       } else {
-        int port_matches;
-        if (wanted->port == CFG_AUTO_PORT) {
-          port_matches = 1;
-        } else {
-          port_matches = (wanted->port == conn->port);
-        }
+        /* Numeric values of old and new port match exactly. */
+        const int port_matches_exact = (wanted->port == conn->port);
+        /* Ports match semantically - either their specific values
+           match exactly, or new port is 'auto'.
+         */
+        const int port_matches = (wanted->port == CFG_AUTO_PORT ||
+                                  port_matches_exact);
+
         if (port_matches && tor_addr_eq(&wanted->addr, &conn->addr)) {
           found_port = wanted;
           break;
         }
+        const int may_need_rebind =
+          port_matches_exact & bool_neq(tor_addr_is_null(&wanted->addr),
```

Comment:
Should this also check whether the addresses are different?  For example: 1.2.3.4 vs 1.2.3.5.  Also for example: [::] vs 0.0.0.0

Reply:
(We should also test for these cases)

Reply:
I think you want '&&' here semantically, even though it happens to be equivalent to '&'.